worker/worker.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2022 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package worker defines abstractions for parallelizing tasks.
+package worker
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// ErrStopped is the error returned when the worker is stopped.
+var ErrStopped = fmt.Errorf("worker is stopped")
+
+// Void is a convenience struct for workers that do not actually return values.
+type Void struct{}
+
+// WorkFunc is a function for executing work.
+type WorkFunc[T any] func() (T, error)
+
+// Worker represents an instance of a worker. It is same for concurrent use, but
+// see function documentation for more specific semantics.
+type Worker[T any] struct {
+	size int64
+	sem  *semaphore.Weighted
+
+	i           int64
+	results     []*result[T]
+	resultsLock sync.Mutex
+
+	stopped uint32
+}
+
+// result is the internal result representation. It is primarily used to
+// maintain results ordering.
+type result[T any] struct {
+	idx    int64
+	result *Result[T]
+}
+
+// Result is the final result returned to the caller.
+type Result[T any] struct {
+	Value T
+	Error error
+}
+
+// New creates a new worker that executes work in parallel, up to the maximum
+// provided concurrency. Work is guaranteed to be executed in the order in which
+// it was enqueued, but is not guaranteed to complete in the order in which it
+// was enqueued (i.e. this is not a pipeline).
+//
+// If the provided concurrency is less than 1, it defaults to the number of CPU
+// cores.
+func New[T any](concurrency int64) *Worker[T] {
+	if concurrency < 1 {
+		concurrency = int64(runtime.NumCPU())
+	}
+	if concurrency < 1 {
+		concurrency = 1
+	}
+
+	return &Worker[T]{
+		size:    concurrency,
+		i:       -1,
+		sem:     semaphore.NewWeighted(concurrency),
+		results: make([]*result[T], 0, concurrency*4),
```

Comment:
The meaning of "4" isn't immediately obvious, could you add a named const or a comment?